File: examples/Sign.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace myApp
{
    using System.Globalization;

    using Org.BouncyCastle.Crypto.Parameters;
    using Org.BouncyCastle.Security;

    public class Signature
    {
        public Signature(string partnerId, string privateKey, string requestUrl, string requestMethod, string payload)
        {
            this.PartnerId = partnerId;
            this.PrivateKey = privateKey;
            this.RequestUrl = requestUrl;
            this.RequestMethod = requestMethod;
            this.Payload = payload;
        }

        public string PartnerId { get; set; }

        public string PrivateKey { get; set; }

        public string RequestUrl { get; set; }

        public string RequestMethod { get; set; }

        public string TimeStamp { get; private set; }

        public string Payload { get; private set; }

        public string GetSignature(string timestamp)
        {
            this.TimeStamp = timestamp;

            var message = this.PartnerId + "\\n" + this.RequestUrl + "\\n" +
                this.RequestMethod.ToLower() + "\\n" + this.TimeStamp + "\\n" + this.Payload;

            RsaKeyParameters rsaKeyParameter;
            try
            {
                byte[] keyBytes = File.ReadAllBytes(@"/Users/ankit/handydev/repos/service-partners-backend/private.der");
                var asymmetricKeyParameter = PrivateKeyFactory.CreateKey(keyBytes);
                rsaKeyParameter = (RsaKeyParameters)asymmetricKeyParameter;
            }
            catch (System.Exception)
            {
                throw new Exception("Unable to load private key");
            }

            var signer = SignerUtilities.GetSigner("SHA256withRSA");
            signer.Init(true, rsaKeyParameter);
            var messageBytes = Encoding.UTF8.GetBytes(message);
            signer.BlockUpdate(messageBytes, 0, messageBytes.Length);
            var signed = signer.GenerateSignature();
            var hashed = Convert.ToBase64String(signed);
            return hashed;
        }

        private static string GetTimestampInMillis()
        {
            DateTime UnixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            DateTime localDateTime, univDateTime;
            localDateTime = DateTime.Now;          
            univDateTime = localDateTime.ToUniversalTime();
            return ((long)(univDateTime - UnixEpoch).TotalMilliseconds/1000).ToString();
        }

        static void Main(string[] args)
        {
            var partnerId = "partner_name";
            var payload = "{\"user\":{\"first_name\":\"John\",\"last_name\":\"Doe\",\"email\":\"johndoe@example.com\",\"phone_number\":\"2125555555\"},\"address\":{\"address1\":\"10 Main Street\",\"address2\":\"Apt 2\",\"city\":\"New York\",\"state\":\"NY\",\"zipcode\":10012,\"country\":\"US\"},\"order\":{\"partner_order_id\":\"12345\",\"products\":[{\"sku\":\"STSP1001\",\"product_info\":{\"product_name\":\"Arm chair photo\"},\"order_details\":{\"quantity\":2,\"delivery_date\":\"2018-04-21T03:50:02.000Z\"}}]}}";
    	      var url = "http://https://partners.services.handy.com/api/v1/orders";
    	      var httpMethod = "post";

            /* You will need to convert your private key to PKCS8 format using following command:
            * openssl pkcs8 -topk8 -inform PEM -outform DER -in private.pem -out private.der -nocrypt
            */
            var privateKeyPath = "test_private_key.der";

            var timestamp = GetTimestampInMillis();
            var signature = new Signature(partnerId, privateKeyPath, url, httpMethod, payload).GetSignature(timestamp);

            Console.WriteLine("Content-Type: application/json");
            Console.WriteLine("HDY-PARTNER-ID: " + partnerId);
            Console.WriteLine("HDY-TIMESTAMP: " + timestamp);
            Console.WriteLine("HDY-SIGNATURE: " + signature);
        }
    }
}

File: examples/Sign.java
import static java.nio.charset.StandardCharsets.UTF_8;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.Signature;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.Base64;

public class Sign {
    
    public static PrivateKey getPrivateKey(String keyPath) throws Exception {
    	byte[] keyBytes = Files.readAllBytes(Paths.get(keyPath));
    	PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
    	KeyFactory kf = KeyFactory.getInstance("RSA");
    	return kf.generatePrivate(spec);
    }

    public static String sign(PrivateKey privateKey, String requestMessage) throws Exception {
    	Signature privateSignature = Signature.getInstance("SHA256withRSA");
        privateSignature.initSign(privateKey);
        privateSignature.update(requestMessage.getBytes(UTF_8));

        byte[] signature = privateSignature.sign();

        return Base64.getEncoder().encodeToString(signature);
    }
    
    public static String requestMessage(String partnerId, String url, String httpMethod, String timestamp, String payload) {
    	StringBuilder sb = new StringBuilder();
    	String separator = "\\n";
        sb.append(partnerId).append(separator);
        sb.append(url).append(separator); 
        sb.append(httpMethod).append(separator); 
        sb.append(timestamp).append(separator); 
        sb.append(payload); 
        return sb.toString();
    }

    public static void main(String... argv) throws Exception {
    	String payload = "{\"user\":{\"first_name\":\"John\",\"last_name\":\"Doe\",\"email\":\"johndoe@example.com\",\"phone_number\":\"2125555555\"},\"address\":{\"address1\":\"10 Main Street\",\"address2\":\"Apt 2\",\"city\":\"New York\",\"state\":\"NY\",\"zipcode\":10012,\"country\":\"US\"},\"order\":{\"partner_order_id\":\"12345\",\"products\":[{\"sku\":\"STSP1001\",\"product_info\":{\"product_name\":\"Arm chair photo\"},\"order_details\":{\"quantity\":2,\"delivery_date\":\"2017-04-17T03:50:02.000Z\"}}]}}";
    	String partnerId = "partner_name";
    	String url = "https://partners.services.handy.com/api/v1/orders";
    	String httpMethod = "post";
    	String timestamp = "" + System.currentTimeMillis() / 1000;

    	String requestMessage = requestMessage(partnerId, url, httpMethod, timestamp, payload);
    	
        /* You will need to convert your private key to PKCS8 format using following command:
         * openssl pkcs8 -topk8 -inform PEM -outform DER -in private.pem -out private.der -nocrypt
         */
    	PrivateKey privateKey = getPrivateKey("test_private_key.der");
        String signature = sign(privateKey, requestMessage);
        System.out.println("Input for signing function: " + requestMessage);
        System.out.println("Signature: " + signature);
        System.out.println("Required headers:");

        System.out.println("Content-Type: application/json");
        System.out.println("HDY-PARTNER-ID: " + partnerId);
        System.out.println("HDY-TIMESTAMP: " + timestamp);
        System.out.println("HDY-SIGNATURE: " + signature);
    }
}
